File: swift/377_CombinationSumIV.swift
class Solution {
    func combinationSum4(_ nums: [Int], _ target: Int) -> Int {
        var cache: [Int : Int] = [Int : Int]()
        
        func countCombinations(_ sum: Int) -> Int {
            if sum == target {
                return 1
            } else if sum > target {
                return 0
            } else if let cacheSum = cache[sum]{
                return cacheSum
            }
            
            var total: Int = 0
            for (i, n) in nums.enumerated() {
                var potentialSum: Int = sum + n
                if cache[potentialSum] == nil {
                    cache[potentialSum] = countCombinations(potentialSum)
                }
                total += cache[potentialSum]!
            }
            return total
        }
        return countCombinations(0)
    }
}

File: java/377_CombinationSumIV.java
class Solution {
    HashMap<Integer, Integer> cache;
    public int combinationSum4(int[] nums, int target) {
        cache = new HashMap<Integer, Integer>();
        return countCombinations(0, nums, target);
    }
    
    public int countCombinations(int sum, int[] nums, int target) {
        if (sum == target) 
            return 1;
        else if (sum > target)
            return 0;
        else if (cache.get(sum) != null)
            return cache.get(sum);
        
        int total = 0;
        for (int n : nums) {
            int currCombination = sum + n;
            if (cache.get(currCombination) == null) 
                cache.put(currCombination, countCombinations(currCombination, nums, target));
            total += cache.get(currCombination);
        }
        return total;
    }
}
